R/conditional-transform.R

```diff
@@ -0,0 +1,101 @@
+#'                                        ds$pet2 == 'Cat' ~ ds$Opinion2,
+#'                                        ds$pet3 == 'Cat' ~ ds$Opinion3)
+#' ds$dot_opinion <- conditionalTransform(ds$pet1 == 'Dog' ~ ds$Opinion1,
+#'                                        ds$pet2 == 'Dog' ~ ds$Opinion2,
+#'                                        ds$pet3 == 'Dog' ~ ds$Opinion3)
+#' }
+#'
+#' @export
+conditionalTransform <- function (..., default_value=NA, type="categorical", categories=NULL) {
+    formulas <- list(...)
+    n <- length(formulas)
+
+    if (n == 0) {
+        halt("no conditions have been supplied")
+    }
+    if (!type %in% c("categorical", "text", "numeric")){
+        halt("type must be either ", dQuote("categorical"), dQuote("text"), " or ", dQuote("numeric"))
+    }
+    if (type != "categorical" & !is.null(categories)){
+        halt("type is not ", dQuote("categorical"), " ignoring ", dQuote("categories"))
+    }
+
+    cases <- vector("list", n)
+    values <- vector("list", n)
+    for (i in seq_len(n)) {
+        formula <- formulas[[i]]
+        if (!inherits(formula, "formula") || length(formula) != 3) {
+            halt("The case provided is not a formula: ", deparseAndFlatten(substitute(list(...))[[i + 1]]))
+        }
+
+        # TODO: switch ot the more lightweight rlang::bare_eval()?
+        cases[[i]] <- eval(formula[[2]])
+        if (!inherits(cases[[i]], "CrunchLogicalExpr")) {
+            halt("The LHS provided is not a CrunchLogicalExpr: ", deparseAndFlatten(substitute(list(...))[[i + 1]][[2]]))
+        }
+        values[[i]] <- eval(formula[[3]])
+    }
+
+    n_rows <- nrow(CrunchDataset(crGET(datasetReference(cases[[1]]))))
+
+    case_indices <- lapply(cases, which)
+
+    # dedpulicate indices, favoring the first observation
+    case_indices <- lapply(seq_along(case_indices), function(i) {
+        setdiff(case_indices[[i]], unlist(case_indices[seq_len(i-1)]))
+    })
+
+    values_to_fill <- Map(function(ind, var) {
+        if (inherits(var, "CrunchVariable")) {
+            # grab the variable contents at inds
+            return(as.vector(var[ind]))
+        } else {
+            # if var isn't a crunchvariable, just return var
+            return(var)
+        }
+    }, ind = case_indices, var = values)
+
+    # setup NAs for as default
+    result <- rep(default_value, n_rows)
+
+    for (i in seq_along(case_indices)) {
+        if (type == "numeric") trans <- as.numeric
+        else trans <- as.character
+        vals <- trans(values_to_fill[[i]])
+        result[case_indices[[i]]] <- vals
+    }
+
+    # check categories, return factor type="categorical"
+    if (type == "categorical") {
+        uni_results <- unique(result)
+        results_not_categories <- !uni_results %in% categories
+        # if categories are supplied and there are any
+        if (is.null(categories)) {
+            result <- factor(result)
+        } else {
```

Comment:
If specifying categories is just about validation, then I'm more against having it as an argument at all. If we're returning a VarDef and we want to let the user specify a particular category order or something, then I could see some value, perhaps. 

Given the current code, you could move L85-86 inside the "else" and only evaluate them if the validation is happening.